sycl/test-e2e/VirtualFunctions/misc/group-barrier.cpp

```diff
@@ -0,0 +1,175 @@
+// REQUIRES: aspect-usm_shared_allocations
+//
+// On CPU it segfaults within the kernel that performs virtual function call.
+// https://github.com/intel/llvm/issues/15080
+// XFAIL: cpu
+// UNSUPPORTED: gpu
+// On GPU this test (its older version which used nd_item instead of group)
+// used to fail with UR_RESULT_ERROR_PROGRAM_LINK_FAILURE.
+// SPIR-V files produced by SYCL_DUMP_IMAGES could be linked just fine (using
+// both llvm-spirv -r + llvm-link and ocloc).
+// Current version hangs and therefore it is marked as unsupported to avoid
+// wasting time in CI and potentially blocking a machine.
+// Reported in https://github.com/intel/llvm/issues/15068
+//
+// This test checks that group operations (barrier in this case) work correctly
+// inside virtual functions.
+//
+// RUN: %{build} -o %t.out %helper-includes
+// RUN: %{run} %t.out
+
+#include <sycl/detail/core.hpp>
+#include <sycl/group_algorithm.hpp>
+#include <sycl/group_barrier.hpp>
+#include <sycl/usm.hpp>
+
+#include "helpers.hpp"
+
+#include <iostream>
+#include <numeric>
+
+namespace oneapi = sycl::ext::oneapi::experimental;
+
+class BaseOp {
+public:
+  SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(oneapi::indirectly_callable)
+  virtual int apply(int *, sycl::group<1>) = 0;
+};
+
+class SumOp : public BaseOp {
+public:
+  SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(oneapi::indirectly_callable)
+  virtual int apply(int *LocalData, sycl::group<1> WG) {
+    LocalData[WG.get_local_id()] += WG.get_local_id();
```

Comment:
This is read/write, but I'm not sure "read" part is really important for this test. Can we change it to write-only (e.g. `g.get_group_linear() + g.get_local_id()`). Then we'd be able to create `/* virtual ? */ int calc_ref_value(auto global_size, auto local_size) { return /* formula */ }`.

That would simplify lines 109-157 a lot, and would also move the reference value compute close to the device code so that they'd fit in a single screen.

Reply:
Good idea, thanks. I've applied that approach in f92ac8537f1584ec1084fb932eb4824d68fde579